docs/source/migration.mdx

```diff
@@ -545,6 +565,49 @@ await startStandaloneServer(server, {
 
 </MultiCodeBlock>
 
+If you want to migrate quickly to Apollo Server 4 *without* altering your data sources, the snippet below replicates Apollo Server 3's `dataSources` behavior with a custom plugin.
+
+<ExpansionPanel title="Click to expand">
+
+```ts
+import { ApolloServerPlugin, BaseContext } from "@apollo/server";
+import type { DataSource } from "apollo-datasource";
+
+type DataSources = Record<string, DataSource>;
+type DataSourcesFn = () => DataSources;
+
+interface ContextWithDataSources extends BaseContext {
+  dataSources?: DataSources;
+}
+
+export const ApolloDataSources = (options: {
+  dataSources: DataSourcesFn,
+}): ApolloServerPlugin<ContextWithDataSources> => ({
+  requestDidStart: async (requestContext) => {
+    const dataSources = options.dataSources();
+    const initializers = Object.values(dataSources).map(async (dataSource) => {
+      if (dataSource.initialize)
+        dataSource.initialize({
+          cache: requestContext.cache,
+          context: requestContext.contextValue,
+        });
+    });
+
+    await Promise.all(initializers);
+
+    requestContext.contextValue.dataSources = dataSources;
+  },
+});
+
+const server = new ApolloServer({
```

Comment:
I do wonder if this will fully work end-to-end if you also add a `context` function, since the type returned by it won't match what's expected in resolvers... or is that "fixed" by the `dataSources` being optional?  might help to extend the example to include a `context` function so folks know what to return from it (and maybe even a resolver that reads context).

Reply:
So this might not be the correct way to pass in the context's type, but I took a stab at making a complete example:
```ts
import { ApolloServerPlugin, BaseContext } from "@apollo/server";
import type { DataSource } from "apollo-datasource";
import UserAPI from "./datasources/users";

type DataSources = Record<string, DataSource>;
type DataSourcesFn = () => DataSources;

interface ContextWithDataSources extends BaseContext {
  dataSources?: DataSources;
}

export const ApolloDataSources = (options: {
  dataSources: DataSourcesFn,
}): ApolloServerPlugin<ContextWithDataSources> => ({
  requestDidStart: async (requestContext) => {
    const dataSources = options.dataSources();
    const initializers = Object.values(dataSources).map(async (dataSource) => {
      if (dataSource.initialize)
        dataSource.initialize({
          cache: requestContext.cache,
          context: requestContext.contextValue,
        });
    });

    await Promise.all(initializers);

    requestContext.contextValue.dataSources = dataSources;
  },
});

interface MyContext {
  token?: string
  dataSources?: {
    userAPI: UserAPI
  }
}

const server = new ApolloServer<MyContext>({
  typeDefs,
  resolvers: {
    Query: {
      books: (_, __, { dataSources, token }) => {
        // Our contextValue has access to our dataSources 
        // and any *other* values we set in our context function 
        return dataSources.userAPI.getUser(token); // highlight-line
      },
    }
  },
  plugins: [ApolloDataSources({
    dataSources: () => {
      return { userAPI: new UserAPI() } // highlight-line
    }
  })]
});

const { url } = await startStandaloneServer(server, {
  async context({ req }) {
    return { token: req.headers.authorization || "" }
  },
});

console.log(`🚀 Server listening at: ${url}`);
```

Reply:
One minor note here: this only works if the tsconfig settings `strictNullChecks` and `strict` are at their admittedly default settings of `false`; it's pretty common to set them to true these days!  With one of them set to true, this won't work:

```
      me: (_, __, { dataSources, token }) => {
        // Our contextValue has access to our dataSources 
        // and any *other* values we set in our context function 
        return dataSources.userAPI.getUser(token); // highlight-line
      },
```

because TypeScript will (correctly!) realize that `dataSources` and `token` might not be defined, and so you'll need to do `dataSources!.userAPI.getUser(token!)` instead.

Reply:
Given that you're not actually showing this in the docs I'll approve, but just be aware of that if you do end up filling out the example somewhere...

Reply:
This is so good to know, thank you for this context! I super appreciate knowing about these TS best practices for the future! 🙇